swarm/network_test.go

```diff
@@ -0,0 +1,577 @@
+func TestSwarmNetwork(t *testing.T) {
+	for _, tc := range []struct {
+		name     string
+		steps    []testSwarmNetworkStep
+		timeout  time.Duration
+		disabled bool
+	}{
+		{
+			name: "10_nodes",
+			steps: []testSwarmNetworkStep{
+				{
+					nodeCount: 10,
+				},
+			},
+			timeout: 45 * time.Second,
+		},
+		{
+			name: "100_nodes",
+			steps: []testSwarmNetworkStep{
+				{
+					nodeCount: 100,
+				},
+			},
+			timeout:  3 * time.Minute,
+			disabled: !*longrunning,
+		},
+		{
+			name: "inc_node_count",
+			steps: []testSwarmNetworkStep{
+				{
+					nodeCount: 2,
+				},
+				{
+					nodeCount: 5,
+				},
+				{
+					nodeCount: 10,
+				},
+			},
+			timeout:  90 * time.Second,
+			disabled: !*longrunning,
+		},
+		{
+			name: "dec_node_count",
+			steps: []testSwarmNetworkStep{
+				{
+					nodeCount: 10,
+				},
+				{
+					nodeCount: 6,
+				},
+				{
+					nodeCount: 3,
+				},
+			},
+			timeout:  90 * time.Second,
+			disabled: !*longrunning,
+		},
+		{
+			name: "dec_inc_node_count",
+			steps: []testSwarmNetworkStep{
+				{
+					nodeCount: 5,
+				},
+				{
+					nodeCount: 3,
+				},
+				{
+					nodeCount: 10,
+				},
+			},
+			timeout: 90 * time.Second,
+		},
+		{
+			name: "inc_dec_node_count",
+			steps: []testSwarmNetworkStep{
+				{
+					nodeCount: 3,
+				},
+				{
+					nodeCount: 5,
+				},
+				{
+					nodeCount: 25,
+				},
+				{
+					nodeCount: 10,
+				},
+				{
+					nodeCount: 4,
+				},
+			},
+			timeout:  5 * time.Minute,
+			disabled: !*longrunning,
+		},
+	} {
+		if tc.disabled {
+			continue
+		}
+		t.Run(tc.name, func(t *testing.T) {
+			testSwarmNetwork(t, tc.timeout, tc.steps...)
+		})
+	}
+}
+
+// testSwarmNetworkStep is the configuration
+// for the state of the simulation network.
+type testSwarmNetworkStep struct {
+	// number of swarm nodes that must be in the Up state
+	nodeCount int
+}
+
+// file represents the file uploaded on a particular node.
+type file struct {
+	key    storage.Key
+	data   string
+	nodeID discover.NodeID
+}
+
+// check represents a reference to a file that is retrieved
+// from a particular node.
+type check struct {
+	key    string
+	nodeID discover.NodeID
+}
+
+// testSwarmNetwork is a helper function used for testing different
+// static and dynamic Swarm network simulations.
+func testSwarmNetwork(t *testing.T, timeout time.Duration, steps ...testSwarmNetworkStep) {
+	dir, err := ioutil.TempDir("", "swarm-network-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	ctx := context.Background()
+	if timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+
+	swarms := make(map[discover.NodeID]*Swarm)
+	files := make([]file, 0)
+
+	services := map[string]adapters.ServiceFunc{
+		"swarm": func(ctx *adapters.ServiceContext) (node.Service, error) {
+			config := api.NewConfig()
+			config.PssEnabled = false
+
+			dir, err := ioutil.TempDir(dir, "node")
```

Comment:
shouldn't there be a `defer os.RemoveAll(dir)`?

Reply:
Actually, this is by design not needed. In order to avoid tracking and removing all nodes directories, they are all under one temp dir which is created here https://github.com/ethersphere/go-ethereum/pull/352/files/159007b743b062e317806d757e220285920e2303#diff-8135ad1202bf3e545e65dafb7002da9cR194, and removed at the end of the test.